crates/network-scanner-net/src/tokio_raw_socket.rs

```diff
@@ -109,3 +89,44 @@ impl TokioRawSocket {
         lock_socket!(self.socket.clone()).set_broadcast(broadcast)
     }
 }
+
+struct RecvFromFuture<'a> {
+    pub socket: Arc<socket2::Socket>,
+    pub buf: &'a mut [MaybeUninit<u8>],
+    pub id: usize,
+    pub handle: AsyncIoHandle,
+}
+
+impl Future for RecvFromFuture<'_> {
+    type Output = std::io::Result<(usize, SockAddr)>;
+
+    fn poll(mut self: std::pin::Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> std::task::Poll<Self::Output> {
+        tracing::trace!("polling recv_from future");
+        let mut buf = vec![MaybeUninit::uninit(); self.buf.len()];
+        match self.socket.recv_from(&mut buf) {
+            Ok(a) => {
+                self.buf.copy_from_slice(&buf);
+                std::task::Poll::Ready(Ok(a))
+            }
+            Err(e) => {
+                if e.kind() == std::io::ErrorKind::WouldBlock {
+                    tracing::trace!("recv_from would block");
+                    let socket = self.socket.clone();
+                    let id = self.id;
+                    if self
+                        .handle
+                        .awake(socket.clone(), Event::readable(id), cx.waker().clone())
+                        .is_err()
+                    {
+                        tracing::warn!("failed to awake socket");
+                        cx.waker().wake_by_ref();
+                    }
```

Comment:
issue: If we fail at registering the interest for this socket, then I think we should return an error instead of starting a busy loop and slowing down the whole application.